File: ruby/elements/tag.rb
require_relative '../utils'

module Elements
  class Tag < Element
    include Utils
    @@locations = Hash.new {|h,k| h[k] = []}
    attr_reader :article, :location, :origin

    def initialize( compiler, article, defn)
      @article   = article
      @location  = defn[0]
      @rendered  = 0
      @origin    = @location.gsub( /\W/, '_').gsub( /_+/, '_').downcase
      if SPECIAL_CHARS =~ @location
        article.error( 'Bad characters in tag')
        @discard = true
      else
        @@locations[@location] << self
      end
    end

    def self.check_all_rendered
      @@locations.each_value do |elements|
        elements.each do |element|
          element.check_rendered
        end
      end
    end

    def check_rendered
      if @rendered == 0
        @article.error( "Location #{@location} not rendered")
      elsif @rendered > 1
        @article.error( "Location #{@location} multiply rendered")
      end
    end

    def date
      @article.date
    end

    def filename
      super + (@origin ? "?origin=#{@origin}" : '')
    end

    def self.find( name)
      @@locations[name]
    end

    def icon
      @article.icon
    end

    def method_missing( name, *args, &block)
      @article.send( name, *args, &block)
    end

    def multiline?
      false
    end

    def page_content?
      false
    end

    def publish?
      return true unless date
      date <= Time.now
    end

    def render
      @rendered += 1
    end

    def set_origin( parents)
      if @origin == 'default'
        @article.error( 'Tag name default not allowed')
      end
      @article.origin( @origin, parents)
    end
  end
end

File: java/src/com/alofmethbin/jadawin/elements/Tag.java
package com.alofmethbin.jadawin.elements;

import com.alofmethbin.jadawin.Article;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Tag extends Element {
    private static Map<String,List<Tag>> tags = new HashMap<>();
    private String tag, origin;
    private int rendered = 0;
    
    public Tag( Article article, List<String> lines) {
        super( article);
        tag    = checkSpecialChars( lines.get(0));
        origin = tag.replaceAll( "\\W", "_").replaceAll( "__", "_");
        
        synchronized ( Tag.class ) {
            find(tag).add( this);
        }
        
        if ( origin.equals( "default") ) {
            error( "Tag name default not allowed");
        }
    }

    public void addOrigin( Article a) {
        article.addOrigin( origin, a);
    }

    public static void checkAllRendered() {
        for (List<Tag> list: tags.values()) {
            for (Tag t: list) {
                if ( t.hasError() ) {continue;}
                if (t.rendered < 1) {
                    t.article.error( "Location " + t.tag + " not rendered");
                } else if (t.rendered > 1) {
                    t.article.error( "Location " + t.tag + " multiply rendered");
                }
            }
        }
    }

    public static List<Tag> find( String tag) {
        List<Tag> list = tags.get(tag);
        if (list == null) {
            tags.put(tag, list = new ArrayList<>());
        }
        return list;
    }

    public LocalDate getDate() {
        return article.getDate();
    }

    public Image getIcon() {
        return article.getIcon();
    }
    
    public String getTitle() {
        return article.getTitle();
    }

    @Override
    public boolean isContent() {
        return false;
    }

    @Override
    public boolean isMultiline() {
        return false;
    }

    public void render() {
        this.rendered ++;
    }
}
